### DISCUSSION TITLE
OnRoundEndAction triggers immediately if inside another event action

### DISCUSSION THREAD

**Discussion by ubertpendragon** [FakeFishGames/Barotrauma#13966] [Code questions and modding support]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [ ] My issue happened while using mods.

### What happened?

OnRoundEndAction doesn't wait for round end and triggers immediately if put inside another event action

2 examples:

**Simple version: checks if player is alive and in submarine on round end. WORKS**

```xml
    <ScriptedEvent identifier="vipnapping2">
          <TagAction criteria="player" tag="player" />
          <OnRoundEndAction>
            <CheckConditionalAction targettag="player" InPlayerSubmarine="true" IsDead="False">
              <Success>          
                <ReputationAction increase="3" identifier="clowns" targettype="Faction" />
                <MoneyAction amount="500" targettag="triggerer_player" />
              </Success>
              <Failure>
              </Failure>
            </CheckConditionalAction>
          </OnRoundEndAction>
    </ScriptedEvent>
```   
    
    
**Now uses CheckDataAction (typical for not repeating events). TRIGGERS AUTOMATICALLY**

```xml
    <ScriptedEvent identifier="vipnapping3">
      <CheckDataAction identifier="vipnappingnorepeat" condition="eq 0" >
        <Failure>
          <!-- does not spawn the event if already done -->
        </Failure>
        <Success>
          <TagAction criteria="player" tag="player" />
          <OnRoundEndAction>
            <CheckConditionalAction targettag="player" InPlayerSubmarine="true" IsDead="False">
              <Success>          
                <ReputationAction increase="3" identifier="clowns" targettype="Faction" />
                <MoneyAction amount="500" targettag="triggerer_player" />
              </Success>
              <Failure>
              </Failure>
            </CheckConditionalAction>
          </OnRoundEndAction>      
        </Success>
      </CheckDataAction>
    </ScriptedEvent>
```

### Reproduction steps

1. load new campaign
2. `triggerevent vipnapping2`
3. Undock
4. `triggerevent vipnapping3`
5. get xp/rep immediately

### Bug prevalence

Happens every time I play

### Single player or multiplayer?

Other

### Version

v1.4.6.0 (Blood in the Water Update, hotfix 2)

### Which operating system did you encounter this bug on?

Windows

**Comment by Regalis11** [comment]
OnRoundEndAction should not be put inside another event action, it must be at the root level. I think semantically the second example is a little strange: you essentially have a block of code that should execute when something happens during the round, and at the same time want it to happen at the end of the round. If we allowed this, I think it could lead to some unintuitive behavior and ambiguities: should the event continue execution through the OnRoundEndAction or wait until the end of the round? When do the actions after OnRoundEndAction get executed, if ever? When is the event considered completed?

**Reply by ubertpendragon** [reply]
It would make logical sense to have `Onroundendaction` execute inside other code, like in my example I was trying to make an event that checks if a NPC is alive at the end of the round, and then issues a reward.

Found workarounds still, I just add an "end" label at the end of the event and skip the OnRoundEndAction, or place it inside another event
